spinnaker-monitoring-daemon/spinnaker-monitoring/meter_specification_handler.py

```diff
@@ -0,0 +1,498 @@
+# Copyright 2018 Google Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+EXPERIMENTAL
+
+This module contains some handlers for manipulating meter specifications.
+Meter specifications are experimental and not yet in use.
+"""
+
+import logging
+import os
+import re
+import yaml
+
+import command_processor
+import http_server
+import spectator_client
+from spectator_handlers import BaseSpectatorCommandHandler
+
+
+METER_SPEC_FILENAME_DECORATOR = '-meter-specification.yml'
+
+def union_keys(*pos_args):
+  """Return a union of all the keys in multiple dicts.
+
+  Args:
+    pos_args: [list of dict]
+  """
+
+  all_keys = set([])
+  for arg in pos_args:
+    if arg:
+      all_keys.update(set(arg.keys()))
+  return all_keys
+
+
+def encode_yaml(yaml_dict):
+  """Return dictionary as yaml string."""
+  return yaml.safe_dump(yaml_dict, allow_unicode=True,
+                        default_flow_style=False,
+                        encoding='utf-8')
+
+
+def compute_meter_diff(original, final):
+  """Log differences between two meter specs."""
+  changes = []
+  for key in union_keys(original, final):
+    original_value = original.get(key)
+    final_value = final.get(key)
+    if original_value == final_value:
+      continue
+    if original_value is None:
+      changes.append('Added %s=%r' % (key, final_value))
+    elif final_value is None:
+      changes.append('Removed %s' % key)
+    elif isinstance(final_value, list):
+      original_set = set(original_value)
+      final_set = set(final_value)
```

Comment:
Should this function be recursive?  That is, could there ever be nested lists more than one level deep?  Currently it only handles one level of nesting.

Reply:
I intentionally didnt for now but since you mentioned it I'll add it in now since I suspect it might go another level to allow individual tags to have attributes.